### PR TITLE
RavenDB-22076 Vector search - memory optimization & quantization fixes & LINQ

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22076

### Additional description

Memory allocation optimizations and fixes in quantization / similarity methods.

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [x] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs [hunk 1]
```diff
@@ -350,7 +350,10 @@ public object CreateVector(string fieldName, object value)
             // - value is textual or array of textual we're treating them as text input
             // - otherwise, we will write as array of numerical values
             var currentIndexingScope = CurrentIndexingScope.Current;
-
+            if (currentIndexingScope != null && currentIndexingScope.Index.IsOnBeforeExecuteIndexing)
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]: What is that about?

Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]: Dictionary training, there is no reason to generate embeddings / create vectors values in training phase.

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]: Can we give it a better name, I couldn't understand the intent there.

Comment (maciejaszyk) [reply src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs]: `DictionaryTraining` is hidden in the Corax implementation of `IndexPersistence.OnBeforeExecuteIndexing([...]);`. To maintain consistency, I would leave it as it is; however, I will add an explanatory comment.

**Review by arekpalinski — APPROVED** [review]

**Review by ppekrol — APPROVED** [review]

**Pull request merged**